Fill all drawers, if user wants that.

SwissalpS opened the pull request:
* adds a toggle in formspec
* if there is no more space for an item, but still empty drawers and toggle is checked, then controller will start filling the next empty drawer it finds.

SwissalpS commented on lua/controller.lua:
```diff
@@ -49,22 +49,25 @@ local pipeworks_loaded = core.get_modpath("pipeworks") and pipeworks
 local digilines_loaded = core.get_modpath("digilines") and digilines
 
 local function controller_formspec(pos)
+	local use_all = core.get_meta(pos):get_int("use_all")
+	if 0 >= use_all then use_all = "false" else use_all = "true" end
```
yes it could, but that is not easy to read

SmallJoker replied:
Alternatively, using builtin functions.
```suggestion
	local use_all = tostring(core.get_meta(pos):get_int("use_all") > 0)`
```

SwissalpS commented on lua/controller.lua:
```diff
@@ -253,49 +258,68 @@ local function controller_get_drawer_index(pos, itemstring)
+		drawer_net_index[name] = controller_index_slot(drawer_pos, visualid)
+
+		-- Set the controller metadata
+		core.get_meta(pos):set_string("drawers_table_index",
+									  core.serialize(drawer_net_index))
```
because the line becomes too long and thus hard to read

OgelGames replied:
There are many other lines that are longer than this, and luacheck doesn't complain about them...

SwissalpS replied:
Lua can handle a lot longer lines than fit in editor window width.

OgelGames replied:
In that case you also need to change lines 69, 164, 169, 357, and 373; they are all longer that this would be. Do you see what my point is?

OgelGames replied:
> In that case you also need to change lines 66, 69, 164, 169, 357, 358, and 373; they are all longer that this would be. Do you see what my point is?

SwissalpS commented on lua/controller.lua:
```diff
@@ -253,49 +258,68 @@ local function controller_get_drawer_index(pos, itemstring)
 	local meta = core.get_meta(pos)
-	local inv = meta:get_inventory()
+	local name = stack:get_name()
+	local use_all = 0 < meta:get_int("use_all")
```
the idea behind that is, that potentially one could set how many to use.
I did not dig deep enough to see how feasible that would be to track how many of which item are being used.

SwissalpS replied:
point has become mute, thanks to using builtin function suggested by SmallJoker

SwissalpS commented on lua/controller.lua:
```diff
@@ -253,49 +258,68 @@ local function controller_get_drawer_index(pos, itemstring)
+			and drawers.drawer_visuals[core.serialize(drawer_pos)]
+		then
+			leftover = drawers.drawer_insert_object(drawer_pos, stack, visualid)
+			if 0 < leftover:get_count() and use_all then
```
nope, static value first is my opinion. If no static value then least vulnerable variable.
some of us think RPN style :D

SmallJoker replied:
And for everybody else it's confusing as hell to read it this way. Too bad.

SwissalpS commented on lua/controller.lua:
```diff
@@ -401,6 +427,12 @@ local function controller_on_receive_fields(pos, formname, fields, sender)
 		meta:set_string("digilineChannel", fields.digilineChannel)
 	end
+	if fields.use_all then
+		local value = 0
+		if "true" == fields.use_all then value = 1 end
+		meta:set_int("use_all", value)
```
I don't like those ternary forms.

OgelGames commented:
Just did some more testing and found that this only works for items inserted from tubes, the `controller_allow_metadata_inventory_put` function needs to be changed to make it work with items inserted from the players' inventory.

SwissalpS commented:
it was never intended to work when adding items manually.
When player is filling up, he is aware and can choose another drawer.
No items go missing when a player is involved so I don't see the need for that situation.
Besides, that would take the fun away of manual organising.

gpcf commented:
I propose an alternative to this: a stack of drawer upgrades that are assigned when a drawer is full by the controller.

SmallJoker commented on lua/controller.lua:
```diff
@@ -253,49 +258,68 @@ local function controller_get_drawer_index(pos, itemstring)
+
+	-- If the drawer is still empty and the drawer entity is loaded, we will
+	-- put the items in the drawer
+	if content.name == "" and drawers.drawer_visuals[core.serialize(drawer_pos)] then
```
`minetest.hash_node_position(pos)` would be more appropriate, and quite a bit faster than serializing.

SwissalpS replied:
thanks for the hint

SwissalpS replied:
there are many such improvements possible. There is even a custom function written to do what math.abs() does.

SmallJoker replied:
:laughing: Oh no. I see there's some more potential, though being out of scope for this PR.

SwissalpS replied:
actually, until I have a rewrite branch up, this is a good place to tell me what can be improved on the mod

SmallJoker commented on lua/controller.lua:
```diff
@@ -253,49 +258,68 @@ local function controller_get_drawer_index(pos, itemstring)
+		local leftover = drawers.drawer_insert_object(drawer_pos, stack, visualid)
+
+		-- Add the item to the drawers table index and set the empty one to nil
+		drawer_net_index["empty"] = nil
```
`"empty"` is a valid item stack name. Potential conflict. How about a numeric index (perhaps `0`) instead, if you don't want to count the keys of this table.

SwissalpS replied:
what about a pseudo item: "drawers:empty" would that work or should we keep short to minimize space used in meta?
I'm not a fan of mixing numerical and string indexes in tables.

SmallJoker replied:
Valid node names are `[%w_]*`, that's equivalent to `[A-z0-9_]`. If you really want to mix two use-cases into a single table, and want to keep the string type as well, you could use an invalid character `;`. But not even this is a safe solution because `minetest.register_node(":NodeOverride-_operator;;,", {...})` is currently accepted as a node name.

SmallJoker commented on init.lua:
```diff
@@ -88,7 +88,7 @@ if core.get_modpath("default") and default then
 		groups = {choppy = 3, oddly_breakable_by_hand = 2},
 		sounds = drawers.WOOD_SOUNDS,
-		drawer_stack_max_factor = 4 * 8, -- normal chest size
+		drawer_stack_max_factor = 32, -- 4 * 8 normal chest size
```
LuaJIT will evaluate this constant expression anyway, and even for Lua it's a single multiplication. This approach is much more error-prone due to potential calculation errors.
I'm not requesting a revert here but would like to point out that such sort of optimizations are not needed.

SwissalpS replied:
thanks, I think it was in context of formspec that this approach was recommended.

OgelGames commented:
> I propose an alternative to this: a stack of drawer upgrades that are assigned when a drawer is full by the controller.

Why not both? There is plenty of room in the controller formspec for both, and drawers can't have unlimited upgrades.

Also, drawer upgrades are not stackable currently, not sure why that is...

SwissalpS commented:
> > I propose an alternative to this: a stack of drawer upgrades that are assigned when a drawer is full by the controller.
> 
> Why not both? There is plenty of room in the controller formspec for both, and drawers can't have unlimited upgrades.
> 
> Also, drawer upgrades are not stackable currently, not sure why that is...

I feel if we made upgrades stackable and also fixed the formspec so shift-clicking would move them as users expect to happen, then we don't need to upgrade the controller too. Maybe I'm missing something - wouldn't be the first time :D

They are possibly not stackable because that is the easiest way to make sure only one per slot is placed in upgrade inventories.

SwissalpS commented:
I have paused working on this issue, and am looking at options to re-write the mod in a more modern and consistent way.
Example I plan to call the node that contains up to four drawers: cabinet, a partition inside a cabinet is a drawer and what is currently called visual, I'm planing to call a tag. (visual wasn't that bad, I still felt it may help to have a shorter name that is more descriptive, since almost everything is a visual in some way). The connectors stay 'trim' and controller remains that too.
If there are any suggestions how to improve this mod from the ground up, now would be a good time to pile them on me.

OgelGames commented:
> They are possibly not stackable because that is the easiest way to make sure only one per slot is placed in upgrade inventories.

Maybe, except that it is also prevented here:
https://github.com/minetest-mods/drawers/blob/f6bca2ade23479f95d51a37053ef0cdf38f23210/lua/api.lua#L149-L164

SwissalpS commented:
> Maybe, except that it is also prevented here:
> 
> https://github.com/minetest-mods/drawers/blob/f6bca2ade23479f95d51a37053ef0cdf38f23210/lua/api.lua#L149-L164

a) could have been added later, maybe original had only one upgrade slot.
b) maybe they are not stackable to be more like what it's original does.

TODO:
- [ ] figure out if there is a good reason upgrades are not stackable

lnjX commented:
@SwissalpS 

> TODO:
> 
> * [ ]  figure out if there is a good reason upgrades are not stackable

The reason for this was that you should only be able to put 5 upgrades into the upgrade slots. However, you could also solve that using other ways (only allowing one of a stack to be inserted). I'm open for making them stackable.

lnjX commented:
I just tested this, but it does not seem to work for me. (I rebuilt the drawer controller and checked out the option.) Is that the reason for the draft label?

SwissalpS commented:
> ... should only be able to put 5 upgrades into the upgrade slots ....
@lnjX Thanks for clarifying that :)

> I just tested this, but it does not seem to work for me. (I rebuilt the drawer controller and checked out the option.) Is that the reason for the draft label?

Yes, I abandoned this attempt and have started a re-write moving code functions away from the entities. So far:
* [x] Entity stores only an id, 1-4, in it's static data and calls a cached handler on player interaction events.
* [x] Cabinet looks up and uses same handler on tube events.
* [x] Slimmed down meta usage, will probably hear from reviewers if this was a good idea or not.
* [x] Trim node added
* [x] Upgrades are stackable and cabinet's formspec allows only one in each slot. (Am thinking of adding experimental option to allow stacks in upgrade slots). Currently am testing with upgrades' max_stack at 5, it's kinda fun but I guess users will want 99
* [x] Drawers can be locked to a specific item even when empty.
* [x] Cleaned up cabinet formspec to show 8/9 columns depending on used mod.
* [x] Added shift-click support to upgrade inventory of cabinets. (only one upgrade per slot)
* [x] Basic manual in/out works with bug fixed that allowed up to int16_max items to be put into empty drawers.
* [x] Basic input from tubes works
* [x] Drawers don't accept unknown items but allow to quickly remove them, should there already be some inside. Also log warning is issued for admins to notice missing definitions.
* [x] replaced stack_max_factor system with a slightly simpler approach. Other mods can still register their own cabinets with more or less space.
I have begun working on controller.
It shouldn't be too long until I have a working test-verison up for review on my fork. It will not yet have migration support at first.

lnjX commented:
Sounds nice so far. Migration support would be essential though. And what's your "simpler approach" for the item count maximum? I like the stack_max thing because it keeps the virtual "size" that an item was supposed to have. If an item is e.g. only stackable to 10 or something, the drawers should not be able to hold much more than a chest could do. Also some mods might have items that are very common and are stackable to much higher values, e.g. fuel units that are stackable to 1000. Having every mod to register its own drawer (possibly in different wood variants) does not sound like a good solution. A mod also can't prevent that there's another drawer registered that can hold more of the item. It is currently not super important of course, because most mods just use the default stack max for everything.

If you've abandoned this PR, I'll close it.

SwissalpS commented:
> If you've abandoned this PR, I'll close it.
Sure:)

I have the feeling that I caused confusion. A drawer containing minegeld can hold more items than a drawer containing cobble.
That has not changed, just the way it is calculated is now more like standard inventories are calculated with slots. Upgrades add slots, not percentage.

Existing code, and my variant both support other mods from adding their own cabinets with their own base-slot-count.

lnjX commented:
Ah okay, sorry then! Basically since there are no different base sizes the current upgrades also add slots (a steel upgrade adds 32 slots / 1 times the base size to a drawer/cabinet). This wasn't very well presented to the user though. Directly seeing the number of new slots / stacks is better I think. :+1: (And is definitely a good idea when there are different base sizes).

SwissalpS commented:
At last I am able to show what I've been working on.
It's not finished but far enough to get the idea of what I'm working on.
[Feel free to add comments and issues over on my rw2020beta0 branch](https://github.com/SwissalpS/drawers/tree/rw2020beta0)
read at least the first couple lines of the readme.md before installing.
